### PR TITLE
Allow verification instead of creation for infrastructure dependencies

### PR DESCRIPTION
**Pull request by iancooper**
Today JustSaying tries to create the required infrastructure for you. Although creation is a no-op for topic, it is not for queues and subscriptions and requires a check to determine if it has already been done. Without a distributed lock, this can lead to race conditions between components. Although topic is a no-op, IaC approaches may prefer to desired state approaches to topics, which may mean we want to create those outside of the JustSaying initialization too.

This PR allows explicit configuration of whether JustSaying should create missing message oriented middleware (MoM) dependencies, validate them, or just call assuming they exist.

it allows for externally-created infrastructure to be identified by name or by Arn.

It tries to support both the existing workflow of creation by default under the principle of "least surprise' and explicitly set other options, though it allows explicit request for creation as well.

### CODE REVIEW

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace JustSaying.Fluent
     /// <summary>
     /// A class representing a builder for a queue subscription. This class cannot be inherited.
     /// </summary>
-    /// <typeparam name="T">
+    /// <typeparam queueName="T">
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:17]:
```suggestion
    /// <typeparam name="T">
```
Probably a find & replace fail here

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 2]
```diff
@@ -27,54 +27,106 @@ internal QueueSubscriptionBuilder()
...
+        /// <summary>
+        /// Configures the queueName of the queue.
+        /// </summary>
+        /// <param queueName="queueName">The queueName of the queue to subscribe to.</param>
         /// <returns>
         /// The current <see cref="QueueSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <exception cref="ArgumentNullException">
-        /// <paramref name="name"/> is <see langword="null"/>.
+        /// <paramref queueName="queueName"/> is <see langword="null"/>.
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:103]:
```suggestion
        /// <paramref name="queueName"/> is <see langword="null"/>.
```

**Comment by iancooper** [comment]
Work outstanding (@slang25 @gkinsman )

Validate By Name (IC)
 -- Validate when created in another account
Validate by Arn
-- Overload on Topic builder property to support a string that is either name or Arn, parse and call explicit parameter version
-- Use Arn instead of name for validation methods above
Assume Middleware
-- Allow option to assume middleware
-- What initialization does this need to do to get Bus working?
Sample
-- Write a sample that uses this new API
-- Show use of environment variables to obtain names/Arns/config style
-- Experiment with switching between options

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 3]
```diff
@@ -95,43 +147,77 @@ public QueueSubscriptionBuilder<T> WithName(string name)
...
         }
 
+        /// <summary>
+        /// Overrides the policy of generating the topic from the type queueName, and allows it to be passed in instead
+        /// </summary>
+        /// <param queueName="topicName">The topic queueName to use</param>
+        /// <param queueName="topicARN">We do not supply a queueName, instead use this externally created Arn instead. Should be used with validate infrastructure only</param>
+        /// <returns></returns>
+        /// <exception cref="InvalidOperationException"></exception>
+        public QueueSubscriptionBuilder<T> WithTopic(string topicName = null, string topicARN = null)
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:170]:
```suggestion
        public QueueSubscriptionBuilder<T> WithTopic(string topicName = null, string topicArn = null)
```
We follow the .NET naming convention for 3 letter initialisms just like the AWS SDK does

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 4]
```diff
@@ -95,43 +147,77 @@ public QueueSubscriptionBuilder<T> WithName(string name)
...
+
+            return this;
+
+        }
+
         /// <summary>
         /// Creates a tag with no value that will be assigned to the SQS queue.
         /// </summary>
-        /// <param name="key">The key for the tag.</param>
+        /// <param queueName="key">The key for the tag.</param>
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:200]:
```suggestion
        /// <param name="key">The key for the tag.</param>
```

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 5]
```diff
@@ -95,43 +147,77 @@ public QueueSubscriptionBuilder<T> WithName(string name)
...
         /// <returns>
         /// The current <see cref="QueueSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <remarks>Tag keys are case-sensitive. A new tag with a key identical to that of an existing one will overwrite it.</remarks>
         /// <exception cref="ArgumentException">
-        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
+        /// <paramref queueName="key"/> is <see langword="null"/> or whitespace.
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:206]:
```suggestion
        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
```

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 6]
```diff
@@ -95,43 +147,77 @@ public QueueSubscriptionBuilder<T> WithName(string name)
...
         /// </exception>
         public QueueSubscriptionBuilder<T> WithTag(string key) => WithTag(key, null);
 
         /// <summary>
         /// Creates a tag with a value that will be assigned to the SQS queue.
         /// </summary>
-        /// <param name="key">The key for the tag.</param>
-        /// <param name="value">The value associated with this tag.</param>
+        /// <param queueName="key">The key for the tag.</param>
+        /// <param queueName="value">The value associated with this tag.</param>
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:214]:
```suggestion
        /// <param name="key">The key for the tag.</param>
        /// <param name="value">The value associated with this tag.</param>
```

File: src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs [hunk 7]
```diff
@@ -95,43 +147,77 @@ public QueueSubscriptionBuilder<T> WithName(string name)
...
         /// <returns>
         /// The current <see cref="QueueSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <remarks>Tag keys are case-sensitive. A new tag with a key identical to that of an existing one will overwrite it.</remarks>
         /// <exception cref="ArgumentException">
-        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
+        /// <paramref queueName="key"/> is <see langword="null"/> or whitespace.
```
Comment (slang25) [inline src/JustSaying/Fluent/QueueSubscriptionBuilder`1.cs:220]:
```suggestion
        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
```

File: src/JustSaying/Fluent/TopicPublicationBuilder`1.cs [hunk 8]
```diff
@@ -114,6 +134,57 @@ public TopicPublicationBuilder<T> WithTag(string key, string value)
...
         }
 
+        /// <summary>
+        /// Overrides the policy of generating the topic from the type name, and allows it to be passed in instead
+        /// </summary>
+        /// <param name="topicName">The topic name to use</param>
+        /// <param name="topicARN">We do not supply a name, instead use this externally created Arn instead. Should be used with validate infrastructure only</param>
+        /// <returns></returns>
+        /// <exception cref="InvalidOperationException"></exception>
+        public TopicPublicationBuilder<T> WithTopic(string topicName = null, string topicARN = null)
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicPublicationBuilder`1.cs:144]:
```suggestion
        public TopicPublicationBuilder<T> WithTopic(string topicName = null, string topicArn = null)
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 9]
```diff
@@ -11,7 +11,7 @@ namespace JustSaying.Fluent
     /// <summary>
     /// A class representing a builder for a topic subscription. This class cannot be inherited.
     /// </summary>
-    /// <typeparam name="T">
+    /// <typeparam topicName="T">
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:14]:
```suggestion
    /// <typeparam name="T">
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 10]
```diff
@@ -24,54 +24,136 @@ internal TopicSubscriptionBuilder()
...
+
+            return this;
+
+        }
+
         /// <summary>
         /// Configures the SNS read configuration.
         /// </summary>
-        /// <param name="configure">A delegate to a method to use to configure SNS reads.</param>
+        /// <param topicName="configure">A delegate to a method to use to configure SNS reads.</param>
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:151]:
```suggestion
        /// <param name="configure">A delegate to a method to use to configure SNS reads.</param>
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 11]
```diff
@@ -24,54 +24,136 @@ internal TopicSubscriptionBuilder()
...
         /// <returns>
         /// The current <see cref="TopicSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <exception cref="ArgumentNullException">
-        /// <paramref name="configure"/> is <see langword="null"/>.
+        /// <paramref topicName="configure"/> is <see langword="null"/>.
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:156]:
```suggestion
        /// <paramref name="configure"/> is <see langword="null"/>.
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 12]
```diff
@@ -92,12 +174,12 @@ public TopicSubscriptionBuilder<T> WithName(string name)
         /// <summary>
         /// Configures the SNS read configuration.
         /// </summary>
-        /// <param name="configure">A delegate to a method to use to configure SNS reads.</param>
+        /// <param topicName="configure">A delegate to a method to use to configure SNS reads.</param>
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:177]:
```suggestion
        /// <param name="configure">A delegate to a method to use to configure SNS reads.</param>
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 13]
```diff
@@ -92,12 +174,12 @@ public TopicSubscriptionBuilder<T> WithName(string name)
...
         /// <returns>
         /// The current <see cref="TopicSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <exception cref="ArgumentNullException">
-        /// <paramref name="configure"/> is <see langword="null"/>.
+        /// <paramref topicName="configure"/> is <see langword="null"/>.
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:182]:
```suggestion
        /// <paramref name="configure"/> is <see langword="null"/>.
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 14]
```diff
@@ -108,27 +190,27 @@ public TopicSubscriptionBuilder<T> WithReadConfiguration(Action<SqsReadConfigura
         /// <summary>
         /// Creates a tag with no value that will be assigned to the SQS queue.
         /// </summary>
-        /// <param name="key">The key for the tag.</param>
+        /// <param topicName="key">The key for the tag.</param>
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:193]:
```suggestion
        /// <param name="key">The key for the tag.</param>
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 15]
```diff
@@ -108,27 +190,27 @@ public TopicSubscriptionBuilder<T> WithReadConfiguration(Action<SqsReadConfigura
...
         /// <returns>
         /// The current <see cref="TopicSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <remarks>Tag keys are case-sensitive. A new tag with a key identical to that of an existing one will overwrite it.</remarks>
         /// <exception cref="ArgumentException">
-        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
+        /// <paramref topicName="key"/> is <see langword="null"/> or whitespace.
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:199]:
```suggestion
        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 16]
```diff
@@ -108,27 +190,27 @@ public TopicSubscriptionBuilder<T> WithReadConfiguration(Action<SqsReadConfigura
...
         /// </exception>
         public TopicSubscriptionBuilder<T> WithTag(string key) => WithTag(key, null);
 
         /// <summary>
         /// Creates a tag with a value that will be assigned to the SQS queue.
         /// </summary>
-        /// <param name="key">The key for the tag.</param>
-        /// <param name="value">The value associated with this tag.</param>
+        /// <param topicName="key">The key for the tag.</param>
+        /// <param topicName="value">The value associated with this tag.</param>
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:207]:
```suggestion
        /// <param name="key">The key for the tag.</param>
        /// <param nane="value">The value associated with this tag.</param>
```

File: src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs [hunk 17]
```diff
@@ -108,27 +190,27 @@ public TopicSubscriptionBuilder<T> WithReadConfiguration(Action<SqsReadConfigura
...
         /// <returns>
         /// The current <see cref="TopicSubscriptionBuilder{T}"/>.
         /// </returns>
         /// <remarks>Tag keys are case-sensitive. A new tag with a key identical to that of an existing one will overwrite it.</remarks>
         /// <exception cref="ArgumentException">
-        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
+        /// <paramref topicName="key"/> is <see langword="null"/> or whitespace.
```
Comment (slang25) [inline src/JustSaying/Fluent/TopicSubscriptionBuilder`1.cs:213]:
```suggestion
        /// <paramref name="key"/> is <see langword="null"/> or whitespace.
```

**Comment by iancooper** [comment]
Picking up: 

Validate by Arn
-- Overload on Topic builder property to support a string that is either name or Arn, parse and call explicit parameter version
-- Use Arn instead of name for validation methods above

**Pull request closed without merging**